ext/datadog_cov/datadog_cov.c

```diff
@@ -161,26 +201,140 @@ static void dd_cov_update_coverage(rb_event_flag_t event, VALUE data, VALUE self
     return;
   }
 
-  char *filename_ptr = RSTRING_PTR(filename);
-  // if the current filename is not located under the root, we skip it
-  if (strncmp(dd_cov_data->root, filename_ptr, dd_cov_data->root_len) != 0)
+  record_impacted_file(dd_cov_data, filename);
+}
+
+// Get source location for a given class name
+static VALUE get_source_location(VALUE klass_name)
+{
+  return rb_funcall(rb_cObject, rb_intern("const_source_location"), 1, klass_name);
+}
+
+// Get source location for a given class name and swallow any exceptions
+static VALUE safely_get_source_location(VALUE klass_name)
+{
+  return rescue_nil(get_source_location, klass_name);
+}
+
+// Executed on RUBY_INTERNAL_EVENT_NEWOBJ event and captures the source file for the
+// allocated object's class.
+static void process_newobj_event(VALUE tracepoint_data, void *data)
+{
+  VALUE self = (VALUE)data;
+  struct dd_cov_data *dd_cov_data;
+  TypedData_Get_Struct(self, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+
+  // careful: this part of tracepoint is running for each allocated object, make it fast
+  rb_trace_arg_t *tracearg = rb_tracearg_from_tracepoint(tracepoint_data);
+  VALUE new_object = rb_tracearg_object(tracearg);
+
+  enum ruby_value_type type = rb_type(new_object);
+  if (type != RUBY_T_OBJECT && type != RUBY_T_STRUCT && type != RUBY_T_DATA)
   {
     return;
   }
```

Comment:
Note that excluding things such as `RUBY_T_HASH` or `RUBY_T_FILE` will also exclude their subclasses. Which may be an explicit decision...

In any case, it may be worth documenting here why others aren't being considered ATM.

Reply:
Mostly for performance reasons I was going to filter for T_OBJECT only, but then I thought that stricter and Data objects have their fair share of usage in typical Rails apps - not sure how often people extend Hash or File in their projects. I will add it in the comment.